cmd/labs/localcache/jsonfile.go

```diff
@@ -0,0 +1,78 @@
+package localcache
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+func NewLocalCache[T any](dir, name string, validity time.Duration) LocalCache[T] {
+	return LocalCache[T]{
+		dir:      dir,
+		name:     name,
+		validity: validity,
+	}
+}
+
+type LocalCache[T any] struct {
+	name     string
+	dir      string
+	validity time.Duration
+	zero     T
+}
+
+func (r *LocalCache[T]) Load(refresh func() (T, error)) (T, error) {
+	cached, err := r.loadCache()
+	if errors.Is(err, fs.ErrNotExist) {
+		return r.writeCache(refresh)
+	} else if err != nil {
+		return r.zero, err
+	} else if time.Since(cached.Refreshed) > r.validity {
+		return r.writeCache(refresh)
+	}
+	return cached.Data, nil
+}
+
+type cached[T any] struct {
+	Refreshed time.Time `json:"refreshed_at"`
+	Data      T         `json:"data"`
+}
+
+func (r *LocalCache[T]) writeCache(refresh func() (T, error)) (T, error) {
```

Comment:
If this fails due to a network error, it should still return the existing cached version (maybe opt in via a bool?).

Can be useful for setups that are not connected to the internet.

Reply:
added url.Error handling